src/ParallelAlgorithms/LinearSolver/Gmres/ElementActions.hpp

```diff
@@ -36,6 +36,111 @@ struct ResidualMonitor;
 namespace LinearSolver {
 namespace gmres_detail {
 
+template <typename FieldsTag>
+struct PrepareSolve {
```

Comment:
I have so far kept the GMRES and CG algorithms separate, not sharing any actions between them but only utility functions like `contribute_to_reduction_observer`. My reasoning is this:
1. The algorithms are complicated enough to understand in this parallelized form without jumping between shared and specialized code.
2. Most actions are sufficiently different that sharing them has little benefit. The `PrepareSolve` actions of GMRES and CG are fairly similar because the two algorithms build on similar ideas, but I'm going to add other parallel linear solvers that I'll use for preconditioning that will do different things in `PrepareSolve`.
These things considered I think trying to be too general with these actions leads to more confusion and complicated code than keeping the actions specialized to each algorithm.

Reply:
Fair enough. We can leave it as-is here, and perhaps revisit if any future utilities also seem easily generalizable with these actions -- I feel factoring is more optional when there are two similar pieces of code than when there are three or more ;).